Resume from latest observed point

ffakenz opened the pull request:
<!-- Describe your change here -->

🔑 **Motivation** 
> Closes #2206 

Currently, the node only updates its `chainState` when a head transition is observed, and uses the latest observed to resume and synchronize with the chain on restart.  
After long periods of inactivity (without head transitions), this state may be stale, negatively impacting catch-up times.  
The goal is to reduce the time required to resynchronize the node with the chain after such periods.

🔄 **Changes**

* The `TickObserved` event schema has changed: the `chainSlot` field has been replaced with `chainPoint`.  
  > This ensures that upon restart, the node can resume from the latest observed tick, even if no head transition has occurred.
* `Greetings` message now contains a new field `currentSlot` that represents the last known chain slot observed by the chain backend. 
  > This is to enhance client reporting on node start/restarts, making it easier to test.
* Chain layer now uses a non-empty list of starting points as a prefix to find a valid intersection when reconnecting to the chain. 
  > A [follow-up issue](https://github.com/cardano-scaling/hydra/issues/2437) tracks further investigation for the case where a chain rollback occurs while the node is offline.

📝 **Notes**

* ⚠️ Offline mode now uses a hardcoded, dummy random block header hash, when simulating activity.

---

<!-- Consider each and tick it off one way or the other -->
* [x] CHANGELOG updated or not needed
* [x] Documentation updated or not needed
* [x] Haddocks updated or not needed
* [x] No new TODOs introduced or explained herafter

ffakenz committed:
extend IsChainState with ChainPointType

> to maintain the newly added field latestKnownChainPoint to ChainStateHistory

ffakenz committed:
refactor Tick to track ChainPoint instead of ChainSlot

ffakenz committed:
replace ChainPointType by ChainPoint from cardano-api

> by moving from chain slots to chain points we made:
> * IsChainState aware of chain points
> * TickObserved to track chain points

ffakenz committed:
report current chain point as part of Greetings

ffakenz committed:
draft e2e scenario

ffakenz committed:
increased the delay when connecting to hydra-node in end-to-end specs

> to avoid exit exceptions during fast restarts.

ffakenz committed:
track latest observed tick and resume node from it

ffakenz committed:
resume ws server from last observed point by chain layer

> not the last recorded by head state

ffakenz committed:
update CHANGELOG

ffakenz committed:
update api spec

ffakenz committed:
Revert "increased the delay when connecting to hydra-node in end-to-end specs"

This reverts commit 81b53932217f368f705e9410f326d31bd27fddc4.

ffakenz committed:
do not wait for node to be in synced

ffakenz committed:
wait for node to be in sync depending on chain backend

ffakenz committed:
do not wait for node to be in sync for membel already bootstrap scenario

ch1bo commented on CHANGELOG.md:
```diff
@@ -18,6 +18,11 @@ changes.
   - See [Issue #2286](https://github.com/cardano-scaling/hydra/issues/2286) and [PR #2290](https://github.com/cardano-scaling/hydra/pull/2290).
 - Fix the `cabalOnly` development environment to allow for `cabal build hydra-node` without `haskell.nix`.
 
+- **BREAKING** Resume node from the latest observed tick when `--start-chain-from` is not set.
```
I commented in the past why this approach may be problematic: https://github.com/cardano-scaling/hydra/pull/2232#pullrequestreview-3214965561

Please respond to this comment how you think the two points are addressed now?

ffakenz replied:
Good catch.

This is my bad, is not the case.
The behavior on startup around `--start-chain-from` has not changed, and the logic remains the same:
```hs
chainPoint <- maybe (queryTip backend) pure $ do
    (max <$> startChainFrom <*> persistedPoint)
      <|> persistedPoint
      <|> startChainFrom
```

ch1bo replied:
I don't mean the `--start-chain-from` bit, but rather the fact that using `Tick` for extending the `ChainStateHistory` is problematic:

> **However**, I'm not sure if this approach is good enough for two reasons:
> 1. The `ChainStateHistory` would become VERY large and this could lead to problems. It was holding one entry per observed head transaction so far, now it holds one per block observed on the chain!
> 2. In presence of chain re-organisations (rollbacks), we would always use the latest known chain state to find an intersection / start observing the chain on startup. If that latest known chain state is as close to the tip as sketched here, it very well may not be on the new longest chain. We must make sure that we try to find intersections with a long enough prefix of `ChainStateHistory` to not "lose our connection". I'm also not sure how the Blockfrost backend deals with this situation.

How are these two things considered now?

ffakenz replied:
Thanks for raising this, good points! 🙏 

I think there might be a small misunderstanding about how `Tick` interacts with `ChainStateHistory` today.

* We **don’t append a new history entry per block**. The history still only grows on `Observation` (i.e. when we actually see a relevant on-chain transaction). On `Tick`, we merely update `latestKnownChainPoint` via `trackLatestKnownChainPoint`, without extending `history`, so the size characteristics stay the same as before.

* For rollbacks, we **first rely on the stored history** (`rollbackHistory`) to rewind to a safe prefix, and only then update `latestKnownChainPoint` accordingly. This ensures we don’t start from a point that’s too close to the tip or no longer on the longest chain.

So the intention is to keep history bounded to meaningful observations, while still tracking the freshest chain point for intersection purposes.

Happy to clarify further if I’m missing an edge case here 🙂

ffakenz committed:
fix CHANGELOG entry

ch1bo commented on hydra-node/src/Hydra/Chain.hs:
```diff
@@ -229,28 +230,38 @@ instance (ArbitraryIsTx tx, IsChainState tx) => Arbitrary (PostTxError tx) where
 data ChainStateHistory tx = UnsafeChainStateHistory
   { history :: NonEmpty (ChainStateType tx)
   , defaultChainState :: ChainStateType tx
+  , latestKnownChainPoint :: ChainPoint
```
Must: not make this cardano specific

Good idea to just store a single point from where we can start next to the history of relevant chain states. However, I believe instead of `ChainPoint`, this should be a `ChainState` (I'd also suggest renaming this to `lastKnown :: ChainStateType tx`).

All usages of `latestKnownChainPoint` have a concrete `tx` instantiated and thus this should be possible.

This also avoids the `SimpleTx` instance to need to import `ChainPoint` and not get polluted with a cardano specific concepts.

ch1bo commented on hydra-node/src/Hydra/Ledger/Simple.hs:
```diff
@@ -103,15 +104,15 @@ instance IsTx SimpleTx where
 -- * Simple chain state
 
-newtype SimpleChainState = SimpleChainState {slot :: ChainSlot}
+newtype SimpleChainState = SimpleChainState {point :: ChainPoint}
```
Must: not use cardano specific concepts in the `Simple` variant of the chain

`ChainPoint` is a Cardano-specific concept and feels wrong to needing it on the `Simple` instance of `IsChainState`

ch1bo replied:
I explained here why this is bad and you responded with a :+1: and resolved the conversation.

This is not resolved as it is **still** using a Cardano `ChainPoint` in the `Simple` instance.

Consequently, the `IsChainState` type class **must not** have a `chainStatePoint` function.

ch1bo commented on hydra-node/src/Hydra/Chain.hs:
```diff
@@ -316,7 +327,7 @@ data ChainEvent tx
     Tick
       { chainTime :: UTCTime
-      , chainSlot :: ChainSlot
+      , chainPoint :: ChainPoint
```
```suggestion
      , tickedChainState :: ChainStateType tx
```

This is consistent with the other constructors and needed if we keep a `lastKnown :: ChainStateType tx` in the `ChainStateHistory tx`

ch1bo commented on hydra-node/src/Hydra/API/WSServer.hs:
```diff
@@ -117,6 +116,7 @@ wsApp env party tracer chain history callback nodeStateP networkInfoP responseCh
             , env
             , networkInfo
             , chainSyncedStatus
+            , atChainPoint = latestKnownChainPoint chainStateHistory
```
Could: make this part of `chainSyncedStatus`

Why is this even needed? The greetings feels like a kitchen sink into which we drop all kinds of things.

ffakenz replied:
This is used for the driver scenario of this feature (`resumeFromLatestKnownPoint`).

ffakenz replied:
I’ll explore the idea of making this part of `chainSyncedStatus`.

ch1bo replied:
Maybe a slot is good enough? That would allow us to the existing `IsChainState` type class API

ch1bo commented on hydra-node/src/Hydra/Chain.hs:
```diff
@@ -229,28 +230,38 @@ instance (ArbitraryIsTx tx, IsChainState tx) => Arbitrary (PostTxError tx) where
+  latestKnownChainPoint =
+    case rolledBack of
+      [] -> ChainPointAtGenesis
+      (cs : _) -> chainStatePoint cs
```
Should: just use `rollbackChainSlot`

or `ChainRolledBack{chainState}` after the other change I suggested

ch1bo reviewed (changes requested):
I like the idea of keeping a single latest state next to the history of chain states we care about.

Besides my (very strong) recommendation of using `ChainStateType tx` directly, I wonder how this behaves if the chain re-organized in between the `hydra-node` (its chain backend) reconnecting:
1. Lets assume the head is open for a while and `head history` is many blocks behind `latestKnownChainPoint`
2. Every `Tick` bumps up `latestKnownChainPoint`, so this tracks the (very) volatile tip of the chain
3. The `hydra-node` is stopped
4. The `cardano-node` we use (or blockfrost uses) sees a small, but common re-organization of the chain by one block -> our `latestKnownChainPoint` is not in the longest chain anymore
5. The `hydra-node` is started

Expectation: the `withDirectChain` backend tries to start chain sync with `latestKnownChainPoint`, but the `chainSyncClient` fails with `IntersectionNotFound`

How is this handled?

ffakenz committed:
do not make chain state history cardano specific

> keep track of latest known chain state instead of chain point

ffakenz committed:
refactor rollback handle

> to work with chain state instead of chain slot

ffakenz committed:
minor fix after rebase

ffakenz committed:
fix start chain from logic

ffakenz committed:
pass a non-empty list of starting points when finding the intersection point during startup

ffakenz commented:
@ch1bo 

Thanks for raising this, that’s a very good scenario to think through.

Based on our discussion, I’ve updated the startup logic so that we no longer try to intersect using only the latest known chain point. Instead, we now pass a **non-empty list of starting points** derived from `ChainStateHistory`, which includes:

  * the latest known chain point, and
  * the bounded history of previous chain states we care about (security depth (k): 2160).

This means that, during a restart, if `the latest known` is ahead of the persisted history and a `re-org` occurred while the `hydra-node` was `offline`, we still expect to find an intersection at one of the points in `history`.

As a follow-up, I’ve opened an issue to explicitly investigate and test this behavior, using the scenario you described: https://github.com/cardano-scaling/hydra/issues/2437

ffakenz committed:
introduce history deph to keep it bounded

ffakenz committed:
do not keep track of lastKnown, instead push new states

> because now history is bounded we can safely use all its points
as candidates to find an intersection

ch1bo commented on hydra-node/test/Hydra/API/ServerSpec.hs:
```diff
@@ -374,17 +379,26 @@ testClient queue semaphore cnx = do
+dummyBaseChainHandle chainStateType =
   Chain
-    { mkChainState = error "unexpected call to mkChainState"
+    { mkChainState = chainStateType
```
Does this mean we expect a call to `mkChainState` now? Why?

ffakenz replied:
This is because [here](https://github.com/cardano-scaling/hydra/pull/2407/files#diff-1a0be0abc025b6d6ed514ed5cf196385a44d3c53abd608531efd65d592188943R119), the API layer now forces evaluation of the node's chain-state to compute `currentSlot` for `Greetings`.

ch1bo replied:
Okay, I have removed that in my further work so we should be able to revert this back to what it was.

ch1bo commented on hydra-node/src/Hydra/Chain.hs:
```diff
@@ -228,29 +229,40 @@ instance (ArbitraryIsTx tx, Arbitrary (ChainStateType tx), IsChainState tx) => A
+  -- ^ The sequence of known chain states, ordered from most recent to oldest.
   , defaultChainState :: ChainStateType tx
+  -- ^ The default chain state to fall back to when rolling back beyond known history.
   }
```
Must: keep a `lastKnown :: ChainStateType tx`

That's the whole point of this PR? The entries in `history` are only ever updated with `pushNewState` and this one is only called on observing a hydra tx. Hence, this is **not** changing the node behavior in that it synchronizes less blocks on startup??

ch1bo replied:
I need to correct myself herer, I overlooked the fact that all `TickObserved` are having `chainState` entries now and we do build a `ChainStateHistory` from them in `aggregateChainStateHistory`.

I think this is correct, but _maybe_ problematic:
- `chainState` is quite big as we keep the whole `spendableUTxO` with it - must check how the memory profile and disk usage behaves if we keep it with `TickObserved`
  - OTOH, using `ChainPoint` directly in the `Chain` api is undesirable as it is a cardano specific type. We could explore a `ChainPointType tx` associated type or a generic type (like `ChainSlot`) .. there was a trade-off I can't remember right now.
- the whole purpose of `ChainStateHistory` especially in comparison to `LocalChainState` is blurry to me and I feel we are mixing two concepts here; the fact I was missing this other code path is evidence to this (ideally we have a better separation of responsibilities and things are done in single place)

ch1bo reviewed (changes requested):
You regressed this into a state where it does not fulfill its original purpose?

Also, you introduced again a leaky abstraction to `IsChainState`, which is not even needed.

ch1bo committed:
Simplify Greetings again to only report on currentSlot

This does not break abstraction and expose cardano specifics on the
websocket server. Also, we can use the NodeState to report on this and
don't need to provide the chain state history (avoids coupling).

ch1bo committed:
Use all of history when selecting a prefix for chain sync

This should retain the original logic where start chain from was used if
it was newer than the last known persisted point.

ch1bo committed:
Avoid leaky abstraction of IsChainState

This does not force the Simple instance to use the Cardano
type (ChainPoint) and prepares usage of the parametric type
ChainPointType in the Hydra.Chain interface.

ch1bo committed:
Switch to only persisting ChainPoint on TickObserved

This should keep the overhead minimal by only adding block header hashes
next to slot numbers. It remains a breaking change though.

ch1bo committed:
Further simplify Chain interface by using ChainPointType

This is deliberately a ChainPointType and not a ChainStateType because
modifications to the state-to-keep-track are only expected upon
observations and rollbacks.

ch1bo committed:
Not limit ChainStateHistory by 2160

The history does not contain all ticked points anymore, but only states
when observations were made an the last known point.

If that is too little, we can aggregate the last x points to try to find
an intersection on demand. See
https://github.com/cardano-scaling/hydra/issues/2437 for more details

ffakenz committed:
fix golden files for greetings and tick observed

ffakenz committed:
resolve fixme comment on event schema missmatch

ffakenz committed:
wip: sketch best effort approach when resolving blockfrost chain points from prefix

ch1bo committed:
Fix again prefix selection

This should be the same semantics as before, but with a list of points
coming from the ChainStateHistory (built from persistence)

ffakenz committed:
revert dummyChainHandle refactor

- removed mkChainState from chain handle
- passing initialChainState via API layer

ch1bo commented on hydra-node/src/Hydra/Chain/Offline.hs:
```diff
@@ -200,10 +199,12 @@ tickForever genesis callback = do
     let timeToSleepUntil = slotNoToUTCTime systemStart slotLength upcomingSlot
     sleepDelay <- diffUTCTime timeToSleepUntil <$> getCurrentTime
     threadDelay $ realToFrac sleepDelay
+    blockHash <- generate (hedgehog genBlockHeaderHash)
```
Could: just use a constant dummy value

This would avoid the added dependency to `quickcheck-hedgehog` and be probably good enough for Offline mode.

ch1bo commented on hydra-node/src/Hydra/HeadLogic.hs:
```diff
@@ -1987,15 +1988,15 @@ aggregateChainStateHistory history = \case
   TxInvalid{} -> history
   LocalStateCleared{} -> history
+  -- FIXME: This makes chain sync starting after rollbacks past the chain state impossible
   Checkpoint nodeState -> initHistory $ getChainState nodeState.headState
```
This is something I noticed when reviewing / refactoring on this PR!

I suspect this could mean that if we happen to rotate the event stream into a `Checkpoint` when the latest chain state is not yet immutable, *and* a re-org happens that leaves us without points on the chain after rollback, we might not be able to chain sync / connect to the node any more.

Maybe worth a new issue? Or could be tackled with #2437. A possible solution could be to never rotate logs within the volatile chain (but this might be annoying to implement). Maybe something more pragmatic does the trick.

ffakenz replied:
Nice catch! Very good observation! 🦅

ffakenz replied:
This definitely needs to be addressed in https://github.com/cardano-scaling/hydra/issues/2437

ch1bo commented on hydra-node/src/Hydra/Node/Run.hs:
```diff
@@ -104,7 +104,7 @@ run opts = do
           -- API
           let apiServerConfig = APIServerConfig{host = apiHost, port = apiPort, tlsCertPath, tlsKeyPath, apiTransactionTimeout}
-          withAPIServer apiServerConfig env stateFile party eventSource (contramap APIServer tracer) chain pparams serverOutputFilter (wireClientInput wetHydraNode) $ \(apiSink, server) -> do
+          withAPIServer apiServerConfig env stateFile party eventSource (contramap APIServer tracer) initialChainState chain pparams serverOutputFilter (wireClientInput wetHydraNode) $ \(apiSink, server) -> do
```
This is preferable over extending the `Chain` handle. An alternative could be to make the projection code monoidal (it's a semigroup now that requires an initial value).

ch1bo commented on hydra-node/test/Hydra/API/HTTPServerSpec.hs:
```diff
@@ -664,15 +665,33 @@ apiServerSpec = do
+          -- NOTE: These lines do NOT represent real state events.
+          -- They are synthetic log entries, only used to mimic the shape of the
+          -- state file so the endpoint can extract the `time` field from
+          -- `HeadInitialized` `stateChanged` events during parsing.
```
:see_no_evil:

ffakenz replied:
Created a follow-up issue to address this: https://github.com/cardano-scaling/hydra/issues/2447

ch1bo commented on hydra-node/test/Hydra/API/ServerSpec.hs:
```diff
@@ -330,7 +330,8 @@ spec =
                     , apiTransactionTimeout = 1000000
                     }
-            withAPIServer @SimpleTx config testEnvironment "~" alice (mockSource []) tracer dummyChainHandle defaultPParams allowEverythingServerOutputFilter noop $ \_ -> do
+                initialChainState = SimpleChainState{slot = ChainSlot 0}
```
I added a `Num` instance to `SimpleChainState` and also `ChainSlot` has one.. so both can be created from literals (via `fromInteger`).

```suggestion
                initialChainState = 0
```

ch1bo commented on hydra-node/test/Hydra/API/ServerSpec.hs:
```diff
@@ -402,7 +402,8 @@ withTestAPIServer ::
 withTestAPIServer port actor eventSource tracer action = do
-  withAPIServer @SimpleTx config testEnvironment "~" actor eventSource tracer dummyChainHandle defaultPParams allowEverythingServerOutputFilter noop action
+  let initialChainState = SimpleChainState{slot = ChainSlot 0}
+  withAPIServer @SimpleTx config testEnvironment "~" actor eventSource tracer initialChainState dummyChainHandle defaultPParams allowEverythingServerOutputFilter noop action
```
```suggestion
  withAPIServer @SimpleTx config testEnvironment "~" actor eventSource tracer 0 dummyChainHandle defaultPParams allowEverythingServerOutputFilter noop action
```

ch1bo commented on hydra-node/test/Hydra/HeadLogicSpec.hs:
```diff
@@ -1381,14 +1381,14 @@ chainEffect postChainTx =
 
--- | Create an observation chain input with chain state at given slot.
-observeTxAtSlot :: Natural -> OnChainTx SimpleTx -> Input SimpleTx
+-- | Create an observation chain input with chain state at given point.
```
```suggestion
-- | Create an observation chain input with chain state at given slot.
```

ch1bo commented on hydra-node/test/Hydra/NodeSpec.hs:
```diff
@@ -339,7 +339,7 @@ spec = parallel $ do
 -- to advance the chain slot and ensure the 'NodeState' is in sync. This is
 -- returning the node to allow for chaining with 'runToCompletion'.
-primeWith :: (MonadSTM m, MonadTime m) => [Input tx] -> HydraNode tx m -> m (HydraNode tx m)
+primeWith :: (MonadSTM m, MonadTime m) => [Input SimpleTx] -> HydraNode SimpleTx m -> m (HydraNode SimpleTx m)
```
Could: try to keep this parameterized over `tx`?

ffakenz replied:
With the switch [from](https://github.com/cardano-scaling/hydra/blob/15be04784c2b90fba8ae5383a38702dbc82be72d/hydra-node/src/Hydra/Chain.hs#L319) `chainSlot` [to](https://github.com/cardano-scaling/hydra/pull/2407/files#diff-4c73f4d0316e43992860895e5e5c41cbeca36294c39715bb6cb979ee041c80aeR370) `chainPoint`, the [code](https://github.com/cardano-scaling/hydra/blame/15be04784c2b90fba8ae5383a38702dbc82be72d/hydra-node/test/Hydra/NodeSpec.hs#L346) that advances the chain:

```hs
Tick now (chainSlot + 1)
````

no longer works, because `ChainPointType tx` has no context-free successor.

Advancing a chain point is tx-specific, so keeping this helper polymorphic would require extra constraints or abstractions.

ch1bo commented on CHANGELOG.md:
```diff
@@ -23,6 +23,11 @@ changes.
 - Add `--unsynced-period` CLI option to configure when the node considers itself
   out of sync with the chain. Defaults to half the contestation period.
 
+- **BREAKING** Resume node from the latest observed tick.
```
Should: provide a more user-centric explanation what that means for them

For example: "Use the last seen block when catching up with the chain after starting the node"

ch1bo reviewed (approved):
I think this looks good now!

Only thing I have not reviewed or tested is the blockfrost backend, but you will figure it out

ffakenz committed:
create SimpleChainState from Num literal

Co-authored-by: Sebastian Nagel <ch1bo@users.noreply.github.com>

ffakenz committed:
add fixme comment on /head-initialization hack

ffakenz committed:
use the same dummy block hash on every tick generated in offline mode

- avoids computing random values over and over

ffakenz committed:
minor enhancement by using num literal as initial chain state

ffakenz committed:
update CHANGELOG

ffakenz committed:
minor comment enhancement

Co-authored-by: Sebastian Nagel <ch1bo@users.noreply.github.com>

ch1bo committed:
Fix dependencies of hydra-node
